geode-apis-compatible-with-redis/src/test/java/org/apache/geode/redis/internal/data/RedisStringTest.java

```diff
@@ -313,6 +317,148 @@ public void setExpirationTimestamp_stores_delta_that_is_stable() throws IOExcept
+    byte[] bytes = {1, 1};
+    RedisString string = new RedisString(bytes);
+    // Index from the end of the byte array, a value of -3 is equivalent to bytes.length - 4
+    assertThat(string.bitpos(1, -3, -3)).isEqualTo(7);
+  }
+
+  @Test
+  public void bitposReturnsNegativeOneWhenBitIsNotFound() {
+    // [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+    byte[] bytes = {0, 0};
+    RedisString string = new RedisString(bytes);
+    assertThat(string.bitpos(1, 0, null)).isEqualTo(-1);
+  }
+
+  @Test
+  public void bitposReturnsNegativeOneWhenBitIsNotFoundAndBitIsZeroAndEndIsSet() {
+    // [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
+    byte[] bytes = {-1, -1};
+    RedisString string = new RedisString(bytes);
+    assertThat(string.bitpos(0, 0, bytes.length)).isEqualTo(-1);
+  }
+
+  @Test
+  public void bitposReturnsValueBitArrayLengthWhenBitIsNotFoundAndBitIsZeroAndEndIsNotSet() {
+    // [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
+    byte[] bytes = {-1, -1};
+    RedisString string = new RedisString(bytes);
+    assertThat(string.bitpos(0, 0, null)).isEqualTo(bytes.length * 8);
+  }
+
+  @Test
+  public void bitposReturnsBitPositionWhenEndIsGreaterThanOrEqualToValueArrayLength() {
+    // [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
+    byte[] bytes = {0, 1};
+    RedisString string = new RedisString(bytes);
+    assertThat(string.bitpos(1, 0, bytes.length)).isEqualTo(15);
+    assertThat(string.bitpos(1, 0, bytes.length + 1)).isEqualTo(15);
+  }
+
+  @Test
+  public void bitposReturnsBitPositionWhenStartAndEndAreGreaterThanOrEqualToValueArrayLengthAndLastByteContainsBit() {
+    // [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
+    byte[] bytes = {0, 1};
+    RedisString string = new RedisString(bytes);
+    assertThat(string.bitpos(1, bytes.length, null)).isEqualTo(15);
+    assertThat(string.bitpos(1, bytes.length + 1, null)).isEqualTo(15);
+  }
+
+  @Test
+  public void bitposReturnsNegativeOneWhenStartAndEndAreGreaterThanOrEqualToValueArrayLengthAndLastByteIsDoesNotContainBit() {
+    // [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0]
+    byte[] bytes = {1, 0};
+    RedisString string = new RedisString(bytes);
+    assertThat(string.bitpos(1, bytes.length, bytes.length)).isEqualTo(-1);
+    assertThat(string.bitpos(1, bytes.length + 1, bytes.length + 1)).isEqualTo(-1);
+  }
+
+  @Test
+  public void bitposReturnsNegativeOneWhenValueIsEmptyArray() {
+    RedisString string = new RedisString(new byte[] {});
+    assertThat(string.bitpos(1, 0, 1)).isEqualTo(-1);
+  }
+
+  @Test
+  public void bitposReturnsNegativeOneWhenEndIsLessThanStart() {
+    RedisString string = new RedisString(new byte[] {0, 1});
+    assertThat(string.bitpos(1, 1, 0)).isEqualTo(-1);
+  }
+
+  @Test
+  public void bitposReturnsNegativeOneWhenBitIsNotZeroOrOne() {
+    RedisString string = new RedisString(new byte[] {0, 1});
+    assertThat(string.bitpos(2, 0, 1)).isEqualTo(-1);
+  }
+
```

Comment:
Where possible, we favor adding these types of tests at the integration level (see `AbstractBitPosIntegrationTest`) so we can make sure our behavior matches Native Redis. Could we add these there instead?

Reply:
Ah, sorry for the mistake. I'll look over the contents of `AbstractBitPosIntegrationTest` and expand the coverage if anything's not there that I added here.

Reply:
I left one tase case in the unit test, as it's testing behaviour specific to the implementation in RedisString that can't be exercised via an integration test.

Reply:
The reason it can't be exercised via an integration test is because we catch it and throw an error in the `BitPosExecutor`:
```
 if (bit != 0 && bit != 1) {
      return RedisResponse.error(ERROR_BIT);
    }
```
So I'm not sure we need a unit test for it... though unit tests are cheap and more tests can't hurt. I guess this goes back to our testing strategy, which we can discuss with the team.